### ISSUE TITLE
Not all the examples in the twisted.application.strports docstring are valid

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10619]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#917|
|Type|defect|
|Created|2005-03-11 19:56:59Z|
|Branch|https://github.com/twisted/twisted/tree/strports-examples-917|

Attachments:

* [strports_examples.3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bd3/bd358eb39360c15b66de4295f1a299ed46740267/2136b51d9a96c86f51f87a587ea56b41913a59ab.patch) (2884 bytes) - added by jesstess on 2010-01-18 01:02:43Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__917 917
type__defect defect
reporter__exarkun exarkun
priority__high high
milestone__ 
branch__branches_strports_examples_917 branches/strports-examples-917
branch_author__jesstess jesstess
status__closed closed
resolution__fixed fixed
component__core core
keywords__documentation documentation
time__1110571019000000 1110571019000000
changetime__1264474772000000 1264474772000000
version__ 
owner__ 
cc__exarkun cc__moshez cc__thijs cc__jesstess
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
In particular, the ssl examples which omit private key and certification
information will not work.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_2"></a>|@thijstriemstra commented|
|-|-|

Where can I find that info?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

There's nothing to find.  The ssl strport type requires a certificate and assumes "server.pem" if one isn't specified.  If the named file isn't found, it fails with an OpenSSL error.  The examples that do include filenames don't work either, unless you happen to have certificate/key files around with those names; also most of the examples fail because `server.Site` requires at least one argument.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra set owner to @thijstriemstra<br>@thijstriemstra set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_6"></a>|@thijstriemstra commented|
|-|-|

Replying to [exarkun](#note_3):
> There's nothing to find.  The ssl strport type requires a certificate and assumes "server.pem" if one isn't specified.  If the named file isn't found, it fails with an OpenSSL error.  The examples that do include filenames don't work either, unless you happen to have certificate/key files around with those names; also most of the examples fail because `server.Site` requires at least one argument.

I'd suggest moving the examples from the docstring to the doc section and make them work. Thoughts?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

moving the examples which would be more complicated than one line to get right sounds like a good idea to me.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_8"></a>|@thijstriemstra set owner to @exarkun<br>@thijstriemstra set status to `new`|
|-|-|

Replying to [exarkun](#note_7):
> moving the examples which would be more complicated than one line to get right sounds like a good idea to me.

This reply confuses me, can you clarify?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to @thijstriemstra|
|-|-|

There's the examples in the docstring which work because they're really simple and there's the examples in the docstring which don't work because they omit required parameters or information.  I meant that the ones that don't work should be moved someplace else where the complexity required to make them work won't be overwhelming.

Of course, given that there are modules like `twisted.protocols.amp` with docstrings as long as your leg, it may also make sense to just make the broken examples work, even if that means making the docstring a lot longer.  That at least avoids adding new files and ensures that someone looking at strports will see the strports documentation, rather than overlooking the fact that there's some docs in another file somewhere - a likely possibility, given the poor high-level organization of a lot of our docs.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_11"></a>|@jesstess removed owner|
|-|-|

5 years seems like plenty of time for this ticket to have been open. I think the examples are going to get more face time living in the code, so attached is a patch that has a toy resource and factory at the top of the docstring that are used by the examples. There is also a note about how SSL examples assume a server.pem if certs aren't given. SSL certificate generation doesn't seem like a topic for the docstring or for strports documentation, and I think people understand that the files being referenced in the SSL examples are something they'd have to create to make the examples work.

If people think this direction is reasonable I'll put the patch in a branch.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_12"></a>|@thijstriemstra set owner to @jesstess|
|-|-|

Can you put this in a branch? The docstring mentions you need "server.pem" but the session shows mycert.pem or mykey.pem, I don't see any references to server.pem, so is this a typo or.. I guess this means as an example it's not helpful as it is cause this confuses me when I read it as a user/newbie. But yea let's get this sorted, as you can see in the history of this ticket's comments, it's not been easy to fix something like this.. ;)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_13"></a>|@jesstess commented|
|-|-|

(In [28043]) Branching to 'strports-examples-917'

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_14"></a>|@jesstess commented|
|-|-|

(In [28044]) Apply strports_examples.3.patch

refs [#10619](https://github.com/twisted/twisted/issues/10619)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_15"></a>|@jesstess commented|
|-|-|

(In [28045]) Reword the examples explanation to clarify what is expected of private
key and certificate files for SSL ports.

refs [#10619](https://github.com/twisted/twisted/issues/10619)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_16"></a>|@jesstess removed owner|
|-|-|

The choice of "certKey" as a named parameter made this a little confusing for me since in my understanding of SSL the certificate is this signed object that proves ones identity and isn't really a key. Am I wrong? Anyway, after reading through the source I've clarified the SSL situation in the examples.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun set owner to @jesstess<br>@exarkun set status to `new`|
|-|-|

> The choice of "certKey" as a named parameter made this a little confusing for me since in my understanding of SSL the certificate is this signed object that proves ones identity and isn't really a key. Am I wrong?

No, you're right.  "certificate" would have been a better name than "certKey".  Alternatively, I probably would have gone with "keyFile" and "certificateFile" in place of "privateKey" and "certKey".  But, oh well.

I like the change, except I think the epytext markup for the new python code and for the examples block itself could be improved slightly (atm the epytext parser thinks it's invalid): a paragraph that ends with "::" and is followed by an indented block (you can see this feature in use in the original markup in this docstring) will have that indented block rendered as preformatted fixed-width text.  This would probably be good for the definitions and the examples.

If you want to see how things get rendered, but you don't want to actually go through the process of rendering it yourself (not too hard, but involves a number of steps and I forget if they're documented anywhere at the moment), you can check stuff into the branch, force a build on the documentation builder, and then visit http://buildbot.twistedmatrix.com/builds/ and you'll see a apidocs-revno.tar.bz2 that corresponds to the revision you built; this contains all the generated api docs).

So just fix that markup issue, then please merge.  Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_19"></a>|@jesstess commented|
|-|-|

(In [28067]) Fix broken epytext markup.

refs [#10619](https://github.com/twisted/twisted/issues/10619)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_20"></a>|@jesstess set status to `closed`|
|-|-|

(In [28068]) Merge strports-examples-917

Author: jesstess
Reviewers: thijs, exarkun
Fixes: [#10619](https://github.com/twisted/twisted/issues/10619)

Turn the strports examples in the module-level docstring into valid or
at least explained examples; comment on the default behavior for SSL
services.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
